keps/sig-node/4216-image-pull-per-runtime-class/README.md

```diff
@@ -0,0 +1,1222 @@
+		return nil, err
+	}
+
+	imagesW := make([]ImagesWrapper, len(imgs))
+	if len(opts) == 0 { 
+		for i, img := range imgs {
+			imagesW[i].Images = NewImage(c, img)
+			imagesW[i].RuntimeHandler = c.runtime
+		}
+	} else {
+		for i, img := range imgs {
+			for runtimeHandler, matcher := range listImgOpts.PlatformMatcherMap {
+				tmp := NewImage(c, img)
+				if checkIfValidPlatforMatcher(ctx, tmp.ContentStore(), tmp.Target(), matcher) {
+					imagesW[i].Images = NewImageWithPlatform(c, img, matcher)
+					imagesW[i].RuntimeHandler = runtimeHandler
+				}
+			}
+			if imagesW[i].Images == nil {
+				return nil, fmt.Errorf("valid platform matcher for image %v not found", img)
+			}
+		}
+	}
+	return imagesW, nil
+}
+
+</pre>
+
+
+### Windows platform matcher optional changes
+
+Each platform has its own platform matcher function defined in https://github.com/containerd/containerd/tree/main/platforms. This is used to decide which manifest to pull from an image index/manifest list. <br> The Windows platform matcher is defined at https://github.com/containerd/containerd/tree/main/platforms/default_windows.go .
+
+Windows platform matcher could optionally be extended to also check for host and UVM compatibility if hyperV isolation is being used based on the matrix described here: https://learn.microsoft.com/en-us/virtualization/windowscontainers/deploy-containers/version-compatibility?tabs=windows-server-2022%2Cwindows-10#windows-server-host-os-compatibility 
+
+### Kubelet changes
+
+Kubelet needs to change how it talks about images from just the image name to (imageName, runtimeclass). As discussed in earlier sections, extending the Image struct in CRI to add a field for runtimeclass will make it easier to extend Kubelet API to include this information. 
+The following are scenarios where kubelet behavior needs to be extended to refer to image as its name and runtime class:
+
+1. Before running a container, kuebelet checks to see whether the required image for the container already exists or not by calling the **EnsureImageExists()** function.
+
+PullImageRequest has a field for ImageSpec which can hold info for the runtime class used to pull this image:
+
+<pre>
+message PullImageRequest {
+    // Spec of the image.
+    ImageSpec image = 1;
+    // Authentication configuration for pulling the image.
+    AuthConfig auth = 2;
+    // Config of the PodSandbox, which is used to pull image in PodSandbox context.
+    PodSandboxConfig sandbox_config = 3;
+}
+</pre>
+
+2. Image status request/response can also make use of the newly added runtime class field in ImageSpec to specify the runtime class being used for the image. If none is specified, containerd will assume that the request is for the default runtime class defined for that platform in containerd.toml
+
+<pre>
+message ImageStatusRequest {
+    // Spec of the image.
+    ImageSpec image = 1;
+    // Verbose indicates whether to return extra information about the image.
+    bool verbose = 2;
```

Comment:
can you use diff syntax in this section so it's easier to see what's being added?

Reply:
There are no new changes being added in this message. I've highlighted ImageSpec field to indicate that this is where the new runtime class field is being added and will be populated. The last paragraph in the section write above this talks about how ImageSpec is being extended to add runtime handler field.